ledger/compactcert.go

```diff
@@ -0,0 +1,155 @@
+	// signatures are still being broadcast, so, continue requiring
+	// 100% votes.
+	//
+	// The first +1 comes from CompactCertWorker.broadcastSigs: it only
+	// broadcasts signatures for round R starting with round R+1, to
+	// ensure nodes have the block for round R already in their ledger,
+	// to check the sig.
+	//
+	// The second +1 comes from the fact that, if we are checking this
+	// acceptable weight to decide whether to allow this transaction in
+	// a block, the transaction was sent out one round ago.
+	offset = offset.SubSaturate(basics.Round(proto.CompactCertRounds/2 + 2))
+	if offset == 0 {
+		return total.ToUint64()
+	}
+
+	// In the next proto.CompactCertRounds/2 blocks, linearly scale
+	// the acceptable weight from 100% to 0%.  If we are outside of
+	// that window, accept any weight.
+	if offset >= basics.Round(proto.CompactCertRounds/2) {
+		return 0
+	}
+
+	w, overflowed := basics.Muldiv(total.ToUint64(), proto.CompactCertRounds/2-uint64(offset), proto.CompactCertRounds/2)
+	if overflowed {
+		// Shouldn't happen, but a safe fallback is to accept a larger cert.
+		logging.Base().Warnf("AcceptableCompactCertWeight(%d, %d, %d, %d) overflow",
+			total, proto.CompactCertRounds, certRound, firstValid)
+		return 0
+	}
+
+	return w
+}
+
+// CompactCertParams computes the parameters for building or verifying
+// a compact cert for block hdr, using voters from block votersHdr.
+func CompactCertParams(votersHdr bookkeeping.BlockHeader, hdr bookkeeping.BlockHeader) (res compactcert.Params, err error) {
+	proto := config.Consensus[votersHdr.CurrentProtocol]
+
+	if proto.CompactCertRounds == 0 {
+		err = fmt.Errorf("compact certs not enabled")
+		return
+	}
+
+	if votersHdr.Round%basics.Round(proto.CompactCertRounds) != 0 {
+		err = fmt.Errorf("votersHdr %d not a multiple of %d",
+			votersHdr.Round, proto.CompactCertRounds)
+		return
+	}
+
+	if hdr.Round != votersHdr.Round+basics.Round(proto.CompactCertRounds) {
+		err = fmt.Errorf("certifying block %d not %d ahead of voters %d",
+			hdr.Round, proto.CompactCertRounds, votersHdr.Round)
+		return
+	}
+
+	totalWeight := votersHdr.CompactCertVotersTotal.ToUint64()
+	provenWeight, overflowed := basics.Muldiv(totalWeight, proto.CompactCertWeightThreshold, 100)
+	if overflowed {
+		err = fmt.Errorf("overflow computing provenWeight[%d]: %d * %d / 100",
+			hdr.Round, totalWeight, proto.CompactCertWeightThreshold)
+		return
+	}
+
+	res = compactcert.Params{
+		Msg:          hdr,
+		ProvenWeight: provenWeight,
+		SigRound:     hdr.Round + 1,
+		SecKQ:        proto.CompactCertSecKQ,
+	}
+	return
+}
+
+// validateCompactCert checks that a compact cert is valid.
+func validateCompactCert(certHdr bookkeeping.BlockHeader, cert compactcert.Cert, votersHdr bookkeeping.BlockHeader, lastCertRnd basics.Round, atRound basics.Round) error {
+	proto := config.Consensus[certHdr.CurrentProtocol]
+
+	if certHdr.Round%basics.Round(proto.CompactCertRounds) != 0 {
```

Comment:
Could you please add a test for `basics.Round(proto.CompactCertRounds)` to ensure it's not zero ?

Reply:
Good idea.  I think it was called only from places that already checked for this, but I like having this function be more defensive.  Pushed.